ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,956 @@
+    private final CopyOnWriteArrayList<BlobAccessInfo> generatedBlobAccessInfos;
+    private final int totalPutCount;
+    private final AtomicInteger currentPutCount;
+    private final int maxParallelRequest;
+    private final int maxFailuresPerPutBatchToStopPuts;
+    private final long sleepTimeBetweenBatchPutsInMs;
+    private final ThreadLocalRandom threadLocalRandom;
+
+    public PutThread(PutGetHelper putHelper, CopyOnWriteArrayList<BlobAccessInfo> generatedBlobAccessInfos,
+        final int totalPutCount, AtomicInteger currentPutCount, int maxParallelRequest,
+        int maxFailuresPerPutBatchToStopPuts, long sleepTimeBetweenBatchPutsInMs, ThreadLocalRandom threadLocalRandom) {
+      this.putHelper = putHelper;
+      this.generatedBlobAccessInfos = generatedBlobAccessInfos;
+      this.totalPutCount = totalPutCount;
+      this.currentPutCount = currentPutCount;
+      this.maxParallelRequest = maxParallelRequest;
+      this.maxFailuresPerPutBatchToStopPuts = maxFailuresPerPutBatchToStopPuts;
+      this.sleepTimeBetweenBatchPutsInMs = sleepTimeBetweenBatchPutsInMs;
+      this.threadLocalRandom = threadLocalRandom;
+    }
+
+    @Override
+    public void run() {
+      while (currentPutCount.get() < totalPutCount) {
+        try {
+          int burstCount = threadLocalRandom.nextInt(maxParallelRequest) + 1;
+          logger.info("PutThread producing " + burstCount + " times ");
+          final CountDownLatch countDownLatch = new CountDownLatch(burstCount);
+          final AtomicInteger failureCount = new AtomicInteger(0);
+          for (int j = 0; j < burstCount; j++) {
+            Callback<Pair<String, byte[]>> callback = new Callback<Pair<String, byte[]>>() {
+              @Override
+              public void onCompletion(Pair<String, byte[]> result, Exception exception) {
+                if (exception == null) {
+                  if (currentPutCount.get() < totalPutCount) {
+                    generatedBlobAccessInfos.add(new BlobAccessInfo(result.getFirst(), result.getSecond()));
+                    currentPutCount.incrementAndGet();
+                  }
+                } else {
+                  logger.error("PutBlob failed with " + exception.getStackTrace());
+                  failureCount.incrementAndGet();
+                }
+                countDownLatch.countDown();
+              }
+            };
+            FutureResult futureResult = putHelper.putBlob(callback);
+          }
+          countDownLatch.await(5, TimeUnit.MINUTES);
+          if (failureCount.get() > maxFailuresPerPutBatchToStopPuts) {
+            logger.error(failureCount.get() + "  failures during this batch, hence exiting the Put Thread");
+            break;
+          }
+          Thread.sleep(threadLocalRandom.nextLong(sleepTimeBetweenBatchPutsInMs));
+        } catch (InterruptedException ex) {
+          logger.error("InterruptedException in putThread " + ex.getStackTrace());
+        }
+      }
+      putComplete.set(true);
```

Comment:
you probably want `putComplete` to be set in a `finally()` - else in case of exceptions you might loop forever in the `GetThread`?